feature: support exact control types in FormGroup and FormArray

sharshuv-quotient opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

From the current documentation:

```ts
import { FormGroup } from '@ngneat/reactive-forms';

const group = new FormGroup<Profile>(...);
const address = group.getControl('name') as FormGroup<Profile['address']>;
const city = group.getControl('address', 'city') as FormControl<string>;
```

**"Note that the return type should still be inferred."** - This PR is fixing this issue.

## What is the new behavior?

I'm enabling passing the type of the control / controls **in addition** to the current ability to pass the value type to `FormGroup` or `FormArray`.
For example, this way I can differ between those two, otherwise identical, Form Groups:
```ts
const group1 = new FormGroup<{
  name: string, // for primitive types, FormControl is inferred, instead of AbstractControl, which is what it was before.
  address: FormControl<{ city: string, street: string }>
}>({
  name: new FormControl('Shachar'),
  address: new FormControl({
    city: 'Tel Aviv',
    street: 'Ibn Gavirol'
  }),
})

const group2 = new FormGroup<{
  name: string, // for primitive types, FormControl is inferred, instead of AbstractControl, which is what it was before.
  address: FormGroup<{ city: string, street: string }>
}>({
  name: new FormControl('Shachar'),
  address: new FormGroup({
    city: new FormControl('Tel Aviv'),
    street: new FormControl('Ibn Gavirol')
  }),
})
```

Now I'll get automatic inference for the controls type:

```ts
const control1 = group1.getControl('address'); // FormControl<{city: string, street: string}> is inferred
const control2 = group1.getControl('address'); // FormGroup<{city: string, street: string}> is inferred
```

## Does this PR introduce a breaking change?

```
[ ] Yes
[X] No
```

Note those changes are completely backward compatible! For evidence, I did not change existing tests and they all pass.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

NetanelBasal commented:
@itayod @danzrou please review.

itayod commented on projects/ngneat/reactive-forms/src/lib/formGroup.ts:
```diff
@@ -48,56 +57,67 @@ export class FormGroup<T extends Obj = any, E extends object = any> extends NgFo
-    validatorOrOpts?: ValidatorOrOpts,
-    asyncValidator?: AsyncValidator
-  ) {
+  constructor(public controls: ControlsOfValue<T>, validatorOrOpts?: ValidatorOrOpts, asyncValidator?: AsyncValidator) {
```
will it still be typed for nested forms that includes `FormGroup` or `FormArray` without passing the `FormGroup` or `FormArray` types? 
for example:
```ts
interface ProfileControls {
  firstName: string;
  lastName: string;
  address: {
    street: string,
    city: string
  } 
}

const profileForm = new FormGroup<ProfileControls>({
  firstName: new FormControl(''),
  lastName: new FormControl(''),
  address : new FormGroup({
    street: new FormControl(''),
    city  : new FormControl('')
  })
});
```

will city and still be typed?

sharshuv-quotient replied:
Yes, in that case the form controls will be interpreted as:
```ts
{
  firstName: FormControl<string>;
  lastName: FormControl<string>;
  address: AbstractControl<{
    street: string;
    city: string;
  }>
}
```

Which means that:
```ts
profileForm.getRawValue().address.city; // string
profileForm.getControl('address'); // AbstractControl<{street: string; city: string}>
profileForm.getControl('address', 'street'); // AbstractControl<string>
```
Exactly as it was before these changes.

itayod commented on projects/ngneat/reactive-forms/src/lib/types.ts:
```diff
@@ -62,11 +62,60 @@ export type KeyValueControls<T extends Obj> = {
+ *
+ * The intermediate type is to solve the issue of T being any, thus assignable to all condition and resulting in the "any" type.
+ * */
+type ControlOfValueWithPotentialUnion<T> = T extends AbstractControl ? T : T extends number | string ? FormControl<T> : AbstractControl<T>
```
what about `boolean`?

sharshuv-quotient replied:
You're right. Should probably add null and undefined too. (for types like `number | null`, for example)
Fixed.

itayod commented on projects/ngneat/reactive-forms/src/lib/types.ts:
```diff
@@ -62,11 +62,60 @@ export type KeyValueControls<T extends Obj> = {
+ *   phone: new FormControl<{num: number, prefix: number}>(),
+ * });
+ * */
+export type FlatControls<T extends Object> = {
```
could you add some more tests and examples for it?
for example, add a test for nested form and use `getControl`

sharshuv-quotient replied:
Add a small test to this, and added this to the documentation.

itayod commented on projects/ngneat/reactive-forms/src/lib/types.ts:
```diff
@@ -62,11 +62,60 @@ export type KeyValueControls<T extends Obj> = {
   [K in keyof T]: T[K] extends FormControl<T[K]>
```
does `KeyValueControls` still in use here? or we can delete it?

sharshuv-quotient replied:
It's not. Removed.

sharshuv-quotient replied:
You're right, it's not being used.
Removed.

itayod commented:
@sharshuv-quotient thanks for the quick reaction.

There is one more thing I would like to ask you to do, could you please run your branch on some big project to make sure there are really not breaking changes?

sharshuv-quotient commented:
Unfortunately I don't have a big project already using this library. 
I've been implementing this because I wanted those changes before I'm starting to use this library in my project.
I believe the tests are good enough, but if you have such a project you can try it your self or if you know of an OpenSource one tell me and I'll try.

itayod commented:
@sharshuv-quotient sure we will also validate it from our side, though it would be nice if you could also try to test your branch inside your project (you could use npm link for that...).

sharshuv-quotient commented:
Yes we can, but I think that's outside of the scope of this PR, since this change can be applied without adding this to the form builder too.

sharshuv-quotient commented:
* Resolved conflicts
* Successfully used this branch in a large project. I should say, however, that it's the first time we I'm this module at all (i.e. I replaced Angular's native module for the first time).

@itayod Are you currently checking this on your project?

itayod commented:
@sharshuv-quotient could you resolve the failing test?

sharshuv-quotient commented:
@itayod Fixed.

NetanelBasal commented:
I'm not sure about this solution. You eliminate the need for casting, but you're also forcing me to override my model interface. In most cases, I'll have my model defined, for example - `Article`, and now I need to "duplicate" it just to get the auto casting.

sharshuv-quotient commented:
I'm not *forcing* you to use a different interface. As I said, this is completely backward compatible.
But if you want your FormGroup to KNOW what kind of controls it has, you have to give it more information. That means a different, more details interface.

I don't see it as a duplicated interface, but as a different interface giving different information. If you want the controls interface to be connected to the value interface in some way (and avoid duplication as much as possible), you can do something like that:
```ts
interface Profile {
  firstName: string;
  lastName: string;
  skills: string[];
  address: {
    street: string;
    city: string;
  };
}

interface ProfileControls extends ControlsOf<Profile> {
  skills: FormArray<string>; // overriding AbstractControl<string[]>
  address: FormGroup<Profile['address']> // overidding AbstractControl<Profile['address']>
}

const fg = new FormGroup<ProfileControls>(...)
```

Coly010 commented:
Adding my two cents. I think the ideal goal is for the type system to understand that if you pass interface T, then it can map each key to a form control, form array or form group based on it's type signature.

```ts
export interface MyInterface {
    key1: string; // Becomes FormControl<string>
   groupKey: { nestedKey: string } // Becomes FormGroup<{ nestedKey: FormControl<string> }>
   arrayKey: string[]; // Becomes FormArray<string>
}
```

We don't exactly want the user to have to create a new interface to define this, we should be able to build this interface for them using the type system.

NetanelBasal commented:
@Coly010 exactly. But there is one exception here. You can use an object or an array with a `FormControl`.

NetanelBasal commented:
A reasonable solution to solve this issue without introducing breaking changes it to create something like:

```ts
const fg = new FormGroup<ControlsOf<Profile>>(...)
```

Coly010 commented:
> @Coly010 exactly. But there is one exception here. You can use an object or an array with a `FormControl`.

In those instances, it may be worth telling the user to provide an override for the keys that don't map based on type signature? 
Or expose decorators that can append metadata to the properties in the interface. The downside of that is the coupling it creates.

NetanelBasal commented:
Yes, but from my experience, it's not too common to pass an object/array to a FormControl, so I'll OK with this.

NetanelBasal commented:
@danzrou any thoughts?

Coly010 commented:
The only use-case i can think of is a multi-select form control. That returns an array of values for one FormControl right?

NetanelBasal commented:
Usually, it will be a custom control accessor that takes an object/array.

sharshuv-quotient commented:
@Coly010 Correct me if I'm wrong, but what you are suggesting here is make the **default** interpretation of the value type regard object as a FormGroup and array as FormArray?
In my opinion this is a bad idea for two reasons:
1. It's a breaking change. Existing code with the `FormGroup<Person>` type which initialized "address" with a FormControl will now break. I think it's better to have the default interpretation ambiguous, than to assume potentially incorrect intentions of the consumer.
2. From my experience it's not too uncommon to use a FormControl with an object type / array type. I have a lot of use-cases in which I'm creating a custom form component (control value accessor) which value is an object / array. I do not think those usecases are as uncommon as you make them seem. 

> A reasonable solution to solve this issue without introducing breaking changes is to create something like:
> 
> ```ts
> const fg = new FormGroup<ControlsOf<Profile>>(...)
> ```

@NetanelBasal You mean here the `ControlsOf` interface is what @Coly010 suggested? I.e. - it's simply a utility interface? If so - we can totally add this, without really changing my current implementation. It's 
a bit similar to the `FlatControls` interface I created to make stuff easier.

NetanelBasal commented:
I mean that `ControlsOf<Profile>` should automatically do what you're doing manually:

```ts
interface ProfileControls extends ControlsOf<Profile> {
  skills: FormArray<string>; // overriding AbstractControl<string[]>
  address: FormGroup<Profile['address']> // overidding AbstractControl<Profile['address']>
}
```

sharshuv-quotient commented:
OK, I'll add it, but I think it should be named `DeepControls` to differ it from the utility interface I have used inside the FormGroup implementation. (which regards objects / arrays as AbstractControls)

igormas commented:
I think that providing an override map to `ControlsOf` as a generic parameter can avoid the burden of creating a dedicated interface in case you want to override some of the controls type.
I.e if you want the address control of the profile FormGroup to be FormControl instead of FormGroup you can achieve it by
```
ControlsOf<Profile, {adress: FormControl< Profile['address']>}> 
```

sharshuv-quotient commented:
> I think that providing an override map to `ControlsOf` as a generic parameter can avoid the burden of creating a dedicated interface in case you want to override some of the controls type.
> I.e if you want the address control of the profile FormGroup to be FormControl instead of FormGroup you can achieve it by
> 
> ```
> ControlsOf<Profile, {address: FormControl< Profile['address']>}> 
> ```

We can add this, as long as the default (`ControlsOf<Profile>` without additional param) is interpreted as in the ambiguous way. (i.e. `ControlsOf<Profile>['address']` is an AbstractControl.) Otherwise it's backward incompatible. 

We can also add `DeepControls` (which will not be used INSIDE the FormGroup implementation, ) as utility interface, with the "override" option:
```ts
FormGroup<DeepControls<Profile, { address: FormControl<Profile['address']> }>
``` 
I should also add that adding this is a good idea because it enables consumer to accomplish the same things in multiple ways, and choose the way that makes most sense to the usecase.

To summarize, a consumer will be able to:
1. Use just `FormGroup<Profile>`, which behaves like it previously did. (backward compatibility)
2. Create a dedicated interface (as I initially suggested) - `FormGroup<ProfileControls>`, which can be created by extending an existing interface.
3. Use one of the following 3 utility interfaces, with the optional "override" type:
  a. **ControlsOf**: primitives are FormControl, every thing else is AbstractControl
Example: `ControlsOf<Profile, {address: FormGroup<Profile['address']}>` - address is `FormGroup` and any other object / array is `AbstractControl`
b. **FlatControls**: every control is FormControl
Example: `FlatControls<Profile, {address: FormGroup<Profile['address']>>` - every control is `FormControl` except address which is a `FormGroup`.
c. **DeepControls**: primitives are FormControl, objects are FormGroup, and arrays are FormArray
Example: `DeepControls<Profile, {address: FormControl<Profile['address']>`

itayod commented:
@sharshuv-quotient IMHO lets start with the `ControlsOf` util that @NetanelBasal suggested, and then rethink about the other utils that you have suggested the `FlatControls` and the `DeepControls` but either way, they should not be included as part of current PR

NetanelBasal commented:
It's starting to get messy or confusing with a lot of options. I really want to keep it straight forward and simple. Can we separate it into multiple PRs? I prefer to start with the initial issue and create the following: (option c)

```ts
const group = new FormGroup<ControlOf<Profile, {address: FormControl<Profile['address']>}>();
```

Coly010 commented:
> It's starting to get messy and confusing with a lot of options. I really want to keep it straight forward and simple. Can we separate it into multiple PRs? I prefer to start with the initial issue and create the following: (option c)
> 
> ```ts
> const group = new FormGroup<ControlsOf<Profile>();
> const group = new FormGroup<ControlsOf<Profile, {address: FormControl<Profile['address']>}>();
> ```

Instead of having to specifically state `{address: FormControl<Profile['address']>}>}`, is there anything that could be done using the `Pick<T, keyof T>` type that might allow us to make that line simply:

`const group = new FormGroup<ControlsOf<Profile, {address: FormControl}>();` 

Something like this: 
```ts
ControlsOf<T, {[R extends keyof T]: FormControl<Pick<T, R>> | FormArray<Pick<T, R>> | FormGroup<Pick<T, R>>}
```

sharshuv-quotient commented:
Hey, I am left a bit confused from all the above chat... If some one can please clarify those things to me so I know how to progress:

- The interface `ControlsOf` is currently used INSIDE The `FormGroup`. If I'm changing this I'm gonna cause backward incompatibility, because `FormGroup<Profile>` will be interpreted differently. So are we keeping it as is? If not what exactly do we want to do with it and with the `FormGroup` implementation? 
- Are we adding the other utility interfaces this PR or in the next one? If we're planning to do it in the next one, what's holding this PR from being merged?

itayod commented:
@ShacharHarshuv 
1. we want to add the `ControlOf` utility as an extra option for users that want to use this "strategy" instead of the current one (which you need to cast the controls), 
as for the usage `ControlsOf`, as you can see above, it will be something like this:
`FormGroup<ControlsOf<Profile>>();`  -  this should cast the `Profile` interface with the same structure but resolve the Control's type as follows:
for T of type Object - `FormGroup<T>`
for T of type Array<T> - `FormArray<T>`
the rest should be resolved as `FormControl<T>`

It would be nice if you can add an ability to override the interface as a second parameter:
`FormGroup<ControlsOf<Profile, {address: FormControl}>()`

2. The other utilities should not be included in this PR. Let's finish the `ControlsOf` and then reobserve the other utils...

danzrou commented:
I'll add my two cents here.

While developing this lib, we did try handling this by providing a type which will checkup the return values (@NetanelBasal `ControlType`).
However that resulted in tons of errors which some of them, as mentioned here, were due to being able to set objects on "single" controls and array as well (and it became even uglier when having deep nested objects).

That said, our main goal was to be able to provide this lib **seamlessly with almost no effort** so users could hook up and integrate it quickly and with minimum of writing.

As most of the solutions provided here are decent, they seem (at least from a first sight) to have a potential of not keeping things plain and simple as intended and adding (maybe indeed minimal) boilerplate for the user.

However, unfortunately, after doing a lot of thinking and also messing around a bit with strict types (#26) I couldn't yet think of a better option than providing the controls type explicitly by the user.

That's why I would go with the most simple solution advised here - a utility type that will resolve the values to their "natural" type: 
1. Object -> `FormGroup`
2. Array -> `FormArray`
3. Primitive/other -> `FormControl`

sharshuv-quotient commented:
Hey, I've added the discussed interface, which can be used optionally. 
I have named it differently though: `DeepControls`, to avoid confusion with `ControlsOf` which is used in the FormGroup implementation (i.e. , explicitly, when you write `FormGroup<Profile>`. This is essential for backward compatibility). I've renamed `ControlsOf` to `AbstractControlOf` to make the difference more clear.

I did not implement the Overrides feature since I believe this should be a separate feature in a separate PR.

sharshuv-quotient commented:
Resolved conflicts. 
Can we merge this?

itayod commented:
@sharshuv-quotient as we wrote above we wanted this pr to include **only** the `controlsOf` utility, could you please rename the utility you have created to `controlsOf` and clean the unrelated code?
Also please add the override functionality so it will be easy to use.

sharshuv-quotient commented:
Done.

itayod commented:
Hi @sharshuv-quotient the `controlsOf` looks good, but as we asked before could you please remove all of the other utils? as we don't want them to be a part of this PR... 
please clean them from your commit and remove them from the docs...

In other words, try focus only on the `ControlsOf` feature and please don't touch anything else :)

itayod commented:
> The other utility you are referring to is `FlatControls`, right? Because `AbstractControls` is required for the FormGroup implementation. I can only make it non-exported if you want but it's essential for everything else to work (both for backward compatibility and both for using the new ControlsOf util)

correct, just remove the `FlatControls` for now and that's it :) thanks!

sharshuv-quotient commented:
Resolved conflicts and remove FlatControls interface as discussed.

sharshuv-quotient commented:
Resolved failing tests

sharshuv-quotient commented:
Hey, can we merge this?

itayod commented:
Hi @sharshuv-quotient sorry for the late response... could you resolve those conflicts and revert the "doc" changes that are not related to this feature?

after you do so I think we can merge it :D

sharshuv-quotient commented:
Resolved conflicts. The doc seemed to be up to date though. Can you please point out what exactly is wrong?

danzrou commented:
LGTM

itayod commented:
@sharshuv-quotient well done!

itayod commented:
@ShacharHarshuv I have noticed there is an issue that seems to be due to this pr.

```ts
interface User {
  id: number;
}
let user: User;
const form = new FormGroup({
  id: new FormControl(user.id)
});
// The type of control should be FormControl<number, any> instead it is AbstractControl<number | FormControl<number, any>>.
const control = form.getControl('id'); // AbstractControl<number | FormControl<number, any>>
```
